### ISSUE TITLE
Alerting: Notification should be sent when state changed from no_data to ok

### ISSUE THREAD

**Issue by flopp999** [grafana/grafana#17780]
**What would you like to be added**:
Email sent when [No Data] is OK again
**Why is this needed**:
Today i get an email if no data is available but not any new email when it is OK again.

I have below condition:
```
WHEN avg () OF query (A, 10m, now) IS OUTSIDE RANGE 10 TO 32

If no data or all values are null
SET STATE TO "No Data"

If execution error or timeout
SET STATE TO "Keep Last State"
```

When there is no data in the response from database, Grafana sets a grey line in graph and I received an email with subject [No Data] ..... and a graph.

When the response do have data again I don't get any email with [OK].

Maybe it would be possible to print out what query/Alias that has triggered the No Data alarm

Grafana 6.2.2 Running on Debian

**Comment by dprokop** [comment]
@flopp999 thanks for reporting. Could you pls share Grafana logs for the described situation? 

Possibly duplicate of https://github.com/grafana/grafana/issues/16496

**Comment by flopp999** [comment]
Is it /var/log/grafana/grafana.log?

Is it possible to tell me what you need in the log file, since I cannot paste the whole log file here due to safety?

**Comment by dprokop** [comment]
See details here for logs location: https://grafana.com/docs/installation/troubleshooting/#logging

Please include only the alerting related parts, meaning the ones that are tagged as `logger=alerting*`

**Comment by flopp999** [comment]
I read the other issue, this is the whole set up in alert tab:

Sorry if I didn't wrote that i do have multi conditions

Rule Name Temperature (Indoor) alert Evaluate every 10m For 0m
Conditions
WHEN
avg () OF query (A, 10m, now) IS OUTSIDE RANGE 10 TO 32
OR
avg () OF query (B, 10m, now) IS OUTSIDE RANGE 10 TO 32
OR
avg () OF query (C, 10m, now) IS OUTSIDE RANGE 10 TO 32
OR
avg ()  OF query (D, 10m, now) IS OUTSIDE RANGE 10 TO 32
OR
avg ()  OF query (E, 10m, now) IS OUTSIDE RANGE 10 TO 32
OR
avg ()  OF query (F, 10m, now) IS OUTSIDE RANGE 10 TO 32

No Data & Error Handling
If no data or all values are null SET STATE TO "NO DATA"
If execution error or timeout SET STATE TO "Keep Last State"

sending mail to multi addresses in one Notification channel, separated with ; (semi colon)

**Comment by flopp999** [comment]
```
t=2019-06-27T05:24:34+0200 lvl=info msg="Alert Rule returned no data" logger=alerting.evalContext ruleId=242 name="Temperature (Indoor) alert" changing state to=no_data
t=2019-06-27T05:24:34+0200 lvl=info msg="New state change" logger=alerting.resultHandler alertId=242 newState=no_data prev state=ok
t=2019-06-27T05:24:34+0200 lvl=info msg=Rendering logger=rendering path="d-solo/000000053/d...t?orgId=9&panelId=7"
t=2019-06-27T05:24:36+0200 lvl=info msg="Sending alert notification to" logger=alerting.notifier.email addresses="[abc@def.com......
```

**Comment by flopp999** [comment]
`t=2019-06-27T05:42:08+0200 lvl=info msg="New state change" logger=alerting.resultHandler alertId=242 newState=ok prev state=no_data`

but after this row these is nothing about sending email

**Comment by flopp999** [comment]
By the way, in Grafana graph I get a green line when data is OK again

**Comment by marefr** [comment]
Think you need to change `If no data or all values are null` to `Alerting`. You can change the defaults for this on a global level, see [documentation](https://grafana.com/docs/installation/configuration/#nodata-or-nullvalues). 

![image](https://user-images.githubusercontent.com/1668778/60249370-ff8db600-98c4-11e9-8393-51eee9ab51b7.png)

Let, me know if you still would like support for no_data -> ok notification and we relabel this as a feature request and reopen this issue.

**Comment by marefr** [comment]
Here's the source code for reference:
https://github.com/grafana/grafana/blob/0a3f977ea2c0ea368bf7ca1d9064a9eb0830be04/pkg/services/alerting/notifiers/base.go#L77-L80

**Comment by flopp999** [comment]
Thanks @marefr 

I have no tried to change `If no data or all values are null` to `Alerting`

I now get an email when no data is present and also an email when data is OK again. Problem is that the first alerting email has subject [Alerting], and there is no text in the email telling that the alert is triggered because No Data is present. This is confusing because then you have no idea why it was triggered.

I think I would like to have this as a feature request.

**Comment by marefr** [comment]
Reopened and re-labeled. Added the priority/awaiting-more-evidence to signal that we need more interest shown in this issue (upvotes via GitHub reactions) for us to accept a pull request solving that or making this a high priority for us to start working on it.

**Comment by AndiKu** [comment]
I have same issue. I need a notification on state change from no_data to ok. I need this for clearing the no_data alert in the target system. Setting the ````If no data or all values are null```` to ````Alerting```` is not an option. I need a different kind of notification in case of no_data. The new code was added with #16905 in v6.2.0-beta1 what is fine for no_data --> pending but not for no_data --> ok.
Because of this I have to set back my instance to v6.1.6 until this is fixed.

**Comment by marefr** [comment]
@AndiKu thanks for the additional information. I'm going to investigate this a bit more, but I think you're right and then we'll try to fix this as soon as possible.

**Comment by AndiKu** [comment]
I was thinking some more about best solution. It might be tricky for following case: ````no_data --> pending --> ok````. The ````no_data --> pending```` will not send a clearing (ok-message) and the ````pending --> ok```` also will not send a clearing. In result the ````no_data```` will never cleared by an ok-message.  
So one option is to remember the ````no_data```` before the ````pending````:  
* ````no_data --> pending --> ok```` send ok message
* ````no_data --> pending --> alerting```` send alerting message
* ````no_data --> pending --> no_data```` no message

Another option is to clear the ````no_data```` directly:
* ````no_data --> pending```` send ok-message (alerting-message might follow after pending time)
* ````no_data --> ok```` send ok-message
* ````no_data --> alerting```` send alerting-message

**Comment by marefr** [comment]
Confirmed that this is a bug that we'll try to prioritize and fix

**Comment by papagian** [comment]
This issue is basically split in 2 separate issues: 

- this one for covering the scenario `no_data` -> `ok` which is already merged with #18920
- issue #18940 for the scenario `no_data` -> `pending` -> `ok` which is still under discussion.
